Instruments/python/r_factor.py

```diff
@@ -1,23 +1,63 @@
 import ROOT
 import argparse
+import json
 
 parser = argparse.ArgumentParser(description='Create bash command')
-parser.add_argument('--ch', required=False, type=str, default= "eTau", help= "channel")
-parser.add_argument('--year', required=False, type=str, default= "2018", help= "year")
-parser.add_argument('--input', required=False, type=str, default= " /eos/home-k/kandroso/cms-it-hh-bbtautau/anaTuples/2020-11-09/", help="Input file name")
-#v_eos_directory = "/eos/home-v/vdamante/merged_trees/"
-#v_eos_directory = os.getenv("merged_trees_dir")
-#k_eos_directory= "/eos/home-k/kandroso/cms-it-hh-bbtautau/anaTuples/2020-11-07/"
-#k_eos_directory = os.getenv("k_eos")
+parser.add_argument('--ch', required=False, type=str, default= "all", help= "channel")
+parser.add_argument('--year', required=False, type=str, default= "all", help= "year")
+parser.add_argument('--unc_sources_group', required=False, type=str, default= "all", help="Input file name")
+
 args = parser.parse_args()
-fileName= args.input+args.year+"_"+args.ch+"_central.root"
-treeName= args.ch
-d = ROOT.RDataFrame(treeName, fileName)
-#def GetRFactor(df):
-not_data = '! is_data'
-d=d.Filter(not_data)
-w_before = d.Define( "num", "all_weights.at(0)").Sum('num')
-w_after = d.Define( "den", "all_weights.at(0)*btag_weight_IterativeFit.at(0)").Sum('den')
-r_factor=w_before.GetValue()/w_after.GetValue()
-print r_factor
-#return r_factor
+channels=[]
+if args.ch!="all":
+    for ch in (args.ch).split(","):
+      channels.append(ch)
+else:
+      channels.append("eTau")
+      channels.append("muTau")
+      channels.append("tauTau")
+years=[]
+if args.year!="all":
+    for year in (args.year).split(","):
+      years.append(year)
+else:
+      years.append("2016")
+      years.append("2017")
+      years.append("2018")
+
+unc_sources=[]
+if args.unc_sources_group!="all":
+    for unc in (args.unc_sources_group).split(","):
+      unc_sources.append(unc)
+else:
+      unc_sources.append("Central")
+      unc_sources.append("LES")
+      unc_sources.append("JES")
+
+files = []
+trees = []
+for i in years:
+    for j in channels:
+        for k in unc_sources:
+            filename = i+"_"+j+"_"+k+".root"
+            files.append(filename)
+            trees.append(j)
+
+central_area = "/mnt/data/Dottorato/anaTuples/2020-11-30/"
+r_factors={}
+for file,tree in zip(files,trees):
+    fileName= central_area+file
+    treeName= tree
+    print fileName,treeName
+    d = ROOT.RDataFrame(treeName, fileName)
+    not_data = '! is_data'
+    d=d.Filter(not_data)
+    w_before = d.Define( "num", "weight").Sum('num')
+    w_after = d.Define( "den", "weight*weight_btag_IterativeFit").Sum('den')
```

Comment:
here you should account the effect of btag unc up/down (in addition to the central value)

Reply:
So it must be defined an r_factorUp and r_facotrDown or in which way these uncertainties must be accounted for?

Reply:
yes. And it better to make it in a loop and store them in a big dictionary that will be saved in a json at the end.